numpy/core/src/multiarray/methods.c

```diff
@@ -2076,10 +2079,15 @@ array_setstate(PyArrayObject *self, PyObject *args)
         }
     }
     else {
-        fa->data = PyDataMem_NEW(PyArray_NBYTES(self));
+        npy_intp num = PyArray_NBYTES(self);
+        int elsize = PyArray_DESCR(self)->elsize;
+        if (num == 0 || elsize == 0) {
+            Py_RETURN_NONE;
+        }
+        fa->data = PyDataMem_NEW(num);
         if (PyArray_DATA(self) == NULL) {
             fa->nd = 0;
-            fa->data = PyDataMem_NEW(PyArray_DESCR(self)->elsize);
+            fa->data = PyDataMem_NEW(elsize);
```

Comment:
I realize this was here before your PR, but it seems strange. What is this supposed to do? We could not allocate `num` bytes of memory, so now we allocate `elsize` instead and raise an error?

Your code looks "correct" but the whole idea looks wrong. Maybe it is a separate issue, this code somehow slipped in during different refactorings?

Reply:
I agree, this is quite odd.
I can add that this NULL check was not covered by the tests even before my changes and I found it difficult to reproduce.

Reply:
opened #13724 to clean up the code not related to this PR